### ISSUE TITLE
Cache in Odoo11 lost model_name key

### ISSUE THREAD

**Issue by elgow** [odoo/odoo#24657]
Impacted versions: 11.0

Steps to reproduce: None. Occurred on long-running server after nighttime period of no activity and with no updates of any kind to the server.

Current behavior: Application module `tidb` failed due to cache losing model key `tidb.case` with the following error:
```
[Wed May 09 16:27:50.293960 2018] [wsgi:error] [pid 20840:tid 139917498210048] ERROR:odoo.http:Exception during JSON request handling.
[Wed May 09 16:27:50.294113 2018] [wsgi:error] [pid 20840:tid 139917498210048] Traceback (most recent call last):
[Wed May 09 16:27:50.294166 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/fields.py", line 944, in __get__
[Wed May 09 16:27:50.294211 2018] [wsgi:error] [pid 20840:tid 139917498210048]     value = record.env.cache.get(record, self)
[Wed May 09 16:27:50.294255 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/api.py", line 960, in get
[Wed May 09 16:27:50.294297 2018] [wsgi:error] [pid 20840:tid 139917498210048]     value = self._data[field][record.id][key]
[Wed May 09 16:27:50.294339 2018] [wsgi:error] [pid 20840:tid 139917498210048] KeyError: <odoo.api.Environment object at 0x7f40fcd7e198>
[Wed May 09 16:27:50.294381 2018] [wsgi:error] [pid 20840:tid 139917498210048]
[Wed May 09 16:27:50.294422 2018] [wsgi:error] [pid 20840:tid 139917498210048] During handling of the above exception, another exception occurred:
[Wed May 09 16:27:50.294513 2018] [wsgi:error] [pid 20840:tid 139917498210048]
[Wed May 09 16:27:50.294565 2018] [wsgi:error] [pid 20840:tid 139917498210048] Traceback (most recent call last):
[Wed May 09 16:27:50.294607 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/http.py", line 646, in _handle_exception
[Wed May 09 16:27:50.294649 2018] [wsgi:error] [pid 20840:tid 139917498210048]     return super(JsonRequest, self)._handle_exception(exception)
[Wed May 09 16:27:50.294690 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/http.py", line 307, in _handle_exception
[Wed May 09 16:27:50.294732 2018] [wsgi:error] [pid 20840:tid 139917498210048]     raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
[Wed May 09 16:27:50.294774 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/tools/pycompat.py", line 87, in reraise
[Wed May 09 16:27:50.294816 2018] [wsgi:error] [pid 20840:tid 139917498210048]     raise value
[Wed May 09 16:27:50.294865 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/http.py", line 683, in dispatch
[Wed May 09 16:27:50.294912 2018] [wsgi:error] [pid 20840:tid 139917498210048]     result = self._call_function(**self.params)
[Wed May 09 16:27:50.294955 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/http.py", line 339, in _call_function
[Wed May 09 16:27:50.294997 2018] [wsgi:error] [pid 20840:tid 139917498210048]     return checked_call(self.db, *args, **kwargs)
[Wed May 09 16:27:50.295038 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/service/model.py", line 97, in wrapper
[Wed May 09 16:27:50.295079 2018] [wsgi:error] [pid 20840:tid 139917498210048]     return f(dbname, *args, **kwargs)
[Wed May 09 16:27:50.295120 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/http.py", line 332, in checked_call
[Wed May 09 16:27:50.295162 2018] [wsgi:error] [pid 20840:tid 139917498210048]     result = self.endpoint(*a, **kw)
[Wed May 09 16:27:50.295202 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/http.py", line 927, in __call__
[Wed May 09 16:27:50.295244 2018] [wsgi:error] [pid 20840:tid 139917498210048]     return self.method(*args, **kw)
[Wed May 09 16:27:50.295313 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/http.py", line 512, in response_wrap
[Wed May 09 16:27:50.295358 2018] [wsgi:error] [pid 20840:tid 139917498210048]     response = f(*args, **kw)
[Wed May 09 16:27:50.295400 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/addons/web/controllers/main.py", line 1200, in load
[Wed May 09 16:27:50.295443 2018] [wsgi:error] [pid 20840:tid 139917498210048]     action = request.env[action_type].browse([action_id]).read()
[Wed May 09 16:27:50.295485 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/addons/base/ir/ir_actions.py", line 194, in read
[Wed May 09 16:27:50.295527 2018] [wsgi:error] [pid 20840:tid 139917498210048]     result = super(IrActionsActWindow, self).read(fields, load=load)
[Wed May 09 16:27:50.295569 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/models.py", line 2520, in read
[Wed May 09 16:27:50.295611 2018] [wsgi:error] [pid 20840:tid 139917498210048]     values[name] = field.convert_to_read(record[name], record, use_name_get)
[Wed May 09 16:27:50.295653 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/models.py", line 4657, in __getitem__
[Wed May 09 16:27:50.295695 2018] [wsgi:error] [pid 20840:tid 139917498210048]     return self._fields[key].__get__(self, type(self))
[Wed May 09 16:27:50.295736 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/fields.py", line 948, in __get__
[Wed May 09 16:27:50.295778 2018] [wsgi:error] [pid 20840:tid 139917498210048]     self.determine_value(record)
[Wed May 09 16:27:50.295828 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/fields.py", line 1059, in determine_value
[Wed May 09 16:27:50.295876 2018] [wsgi:error] [pid 20840:tid 139917498210048]     self.compute_value(recs)
[Wed May 09 16:27:50.295919 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/fields.py", line 1015, in compute_value
[Wed May 09 16:27:50.295961 2018] [wsgi:error] [pid 20840:tid 139917498210048]     self._compute_value(records)
[Wed May 09 16:27:50.296002 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/fields.py", line 1006, in _compute_value
[Wed May 09 16:27:50.296044 2018] [wsgi:error] [pid 20840:tid 139917498210048]     getattr(records, self.compute)()
[Wed May 09 16:27:50.296085 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/addons/base/ir/ir_actions.py", line 154, in _compute_search_view
[Wed May 09 16:27:50.296127 2018] [wsgi:error] [pid 20840:tid 139917498210048]     fvg = self.env[act.res_model].fields_view_get(act.search_view_id.id, 'search')
[Wed May 09 16:27:50.296169 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/api.py", line 760, in __getitem__
[Wed May 09 16:27:50.296249 2018] [wsgi:error] [pid 20840:tid 139917498210048]     return self.registry[model_name]._browse((), self)
[Wed May 09 16:27:50.296301 2018] [wsgi:error] [pid 20840:tid 139917498210048]   File "/home/odoo11/odoo/odoo/modules/registry.py", line 181, in __getitem__
[Wed May 09 16:27:50.296386 2018] [wsgi:error] [pid 20840:tid 139917498210048]     return self.models[model_name]
[Wed May 09 16:27:50.296450 2018] [wsgi:error] [pid 20840:tid 139917498210048] KeyError: 'tidb.case'
```
Restart of apache server hosting Odoo as a WSGI app fixed the problem.

Expected behavior: Cache should not spontaneously lose key for models that exist

Video/Screenshot link (optional):

**Comment by pedrobaeza** [comment]
Sorry, but the error you are mentioning corresponds to a custom module, so the problem should be there in your code. This issue tracker is for notifying problems with standard Odoo code, and thus I'm closing this.

Try to share your question, doubt or problem in the official Odoo forums (https://www.odoo.com/forum/help-1) or on the mailing lists (https://www.odoo.com/groups) to see if someone can help you.

**Comment by elgow** [comment]
If you look through the stack trace do you see any calls to non-Odoo code? The model for which the key went missing had been running stably for months and made no changes to its registered models. This is clearly a failure of Odoo code, not of any custom code. 

Is it Odoo's policy to ignore all issues where any non-Odoo module has been installed? That would seem to limit the value of the addon module marketplace, no?

**Comment by pedrobaeza** [comment]
`tidb.case` is not an Odoo model.

**Disclaimer**: I'm not an Odoo employee, but a community member that voluntary contributes to triage / classify / solve issues. Take also into account that there's no obligation to answer nor solving things put in Odoo Community issue tracker. And thus, please consider this as a best effort initiative.

**Comment by elgow** [comment]
`tidb.case` is the name of a model that is, or was, registered in the model registry. The registry handles it exactly as it handles any other model, including Odoo models.  I had hoped that Odoo would be concerned about this apparent problem with the model registry in their latest version.

My immediate problem was fixed by restarting Apache, so I'm not seeking help in solving it.  If this is a problem with the Odoo model cache, as it seems to be, it would be in their interest to understand and fix it before it causes spontaneous failures for other Odoo 11 users. Closing it w/o any apparent consideration of the fact that the entire stack trace is Odoo code precludes gaining anything from this issue report.  I appreciate your volunteering to triage issues, but simply closing them for spurious reasons may not be the best approach. 

I've posted the trace to the forum to see if other Odoo 11 users are seeing spontaneous key loss in their model registries.

**Comment by Yenthe666** [comment]
Hi @elgow,

Do you happen to have a reproducible scenario, steps to trigger this or a custom module that demonstrates what you believe is an issue? If we can reproduce it I'm more than happy to reopen this and look into it further. For now it is hard for us to do anything with the bug report as there are no steps or ways to trigger this. Pedro had closed this a bit fast maybe but I don't see an easy way to verify this and it is custom code, so I can understand his reasoning although it may look a bit harsh.


Just let us know and then we can have a second look.

**Comment by pedrobaeza** [comment]
As you already know, I prefer to have the "house clean". As we can reopen the issue in any moment, if @elgow provides more information, I would be glad to track it and get it solved. Meanwhile, there's no sense on keeping it opened.

**Comment by pedrobaeza** [comment]
@elbow the issue has been closed because there's no error on Odoo itself.

**Comment by gdgellatly** [comment]
@elgow @pedrobaeza @Yenthe666 1 easy way to reproduce is to extend a model and not depend on the parent module in the manifest.  All goes well in development, keeps going well, works in production, then one day when the server restarts for whatever reason the graph order for module loading is different and boom, you get this error.  I've done this to myself a few times, usually when I have a seperate UI customisation or reports module.  

But in any case the report is misleading, there is no key that has gone spontaneously missing.  It is a key that is not yet loaded on restart, and usually (at least for me) this is because dependencies are incorrectly defined.

**Comment by elgow** [comment]
@gdgellatly is quite right, my original description of the problem was wrong, and what he describes fits perfectly with what happened to my module. I will check my manifest to ensure that all parent modules are called out as dependencies. Many thanks.

**Comment by elgow** [comment]
@gdgellatly @Yenthe666 @pedrobaeza I've checked my dependencies and I have not yet found any missing. The only model that I inherit from is `mail.thread` and `mail` is called out in my module dependencies. 

I do have one theory as to what may be happening. I have my models defined in two files, one of which has fields that reference models from the other file. If they were loaded in reverse order then that may cause the problem. AFAIK there is no way that I can control the order in which Odoo loads those files. If I'm wrong about this then I'd definitely appreciate learning how to do it. 

If load order for files in the `models/` directory is indeterminate and it is essential to ensure that there are no cross-references between files then this would be a good thing to document in [Building a Module](https://www.odoo.com/documentation/11.0/howtos/backend.html).

**Comment by Yenthe666** [comment]
@gdgellatly,

Thanks a lot for the insight and info. I think a new report about this would be great, with as much info as possible. If this is really the case then it deserves an open and clean bug report.

**Comment by gdgellatly** [comment]
@elgow in my experience my errors have almost always been either a: in my xml, or b: in my python code accessing an xml reference or another model for which I haven't configured a dependency.  But thats just me, and where I typically make mistakes.

In your case, you are looking in the wrong place.  The problem isn't the module bringing tidb.case, it is modules that ever access it (hence why my errors are typically in UI and Report modules).  Its more than strict code inheritance, its really any dependency in any code on that model.

I've never had an issue with model loading order within a module, at least not since v6, when it was an occassional issue and we used to have to worry about even the order we created our models in a file.
